### ISSUE TITLE
fsfsdf

### ISSUE THREAD

**Issue by nikoleta284** [nikoleta284/Elektrik#1]
import json
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE, RandomOverSampler
import plotly.express as px
from google.colab import files

# Загрузка данных
data = json.loads("""[
    {"accountId":1497,"isCommercial":true,"address":"Краснодарский край, р-н Мостовский, пгт Мостовской, ул Колхозная, д. 37 1","buildingType":"Частный","roomsCount":1,"residentsCount":1,"consumption":{"1":3484,"2":2824,"3":3035,"4":3597,"5":2664,"6":3874,"7":3270,"8":3714,"9":3124,"10":2981,"11":2923,"12":2216}},
    {"accountId":1509,"isCommercial":true,"address":"Краснодарский край, р-н Усть-Лабинский, х Братский, ул Ленина, д. 49 А","buildingType":"Частный","roomsCount":1,"residentsCount":1,"consumption":{"1":3756,"2":1580,"3":3191,"4":2931,"5":793,"6":721,"7":680,"8":581,"9":426,"10":440,"11":1206,"12":2377}},
    {"accountId":1674,"isCommercial":true,"address":"Краснодарский край, р-н Тбилисский, ст-ца Тбилисская, ул Хуторская, д. 31 ","buildingType":"Частный","roomsCount":3,"residentsCount":2,"totalArea":80.50,"consumption":{"1":1543,"2":1075,"3":2344,"4":1125,"5":1045,"6":914,"7":1731,"8":1867,"9":977,"10":1178,"11":1494,"12":1495}},
    {"accountId":1955,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, с Сукко, ул Кооперативная, д. 2 ","buildingType":"Частный","roomsCount":5,"residentsCount":1,"totalArea":908.00,"consumption":{"1":5564,"2":6201,"3":5364,"4":4031,"5":5452,"10":590,"11":4089,"12":5408}},
    {"accountId":1960,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, село Витязево, проезд Красивый, д. 3 ","buildingType":"Частный","roomsCount":3,"residentsCount":3,"consumption":{"1":631,"2":616,"3":439,"4":562,"5":4723,"6":9719,"7":16720,"8":20134,"9":11680,"10":3270,"11":1215,"12":726}},
    {"accountId":1968,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, село Витязево, пер Благовещенский, д. 14 ","buildingType":"Частный","roomsCount":15,"residentsCount":3,"consumption":{"1":320,"2":460,"3":380,"4":480,"5":340,"6":4920,"7":12660,"8":14600,"9":7520,"10":980,"11":820,"12":720}},
    {"accountId":1969,"isCommercial":true,"address":"Краснодарский край, р-н Крымский, х Веселый (Мерчанское с/п), ул Пушкина, д. 1 ","buildingType":"Частный","roomsCount":2,"residentsCount":1,"totalArea":33.80,"consumption":{"1":125,"2":125,"3":125,"4":125,"5":125,"6":125,"7":125,"8":125,"9":125,"10":125,"11":125,"12":125}},
    {"accountId":1971,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, село Витязево, пер Благовещенский, д. 6 , кв. 1 ","buildingType":"Частный","roomsCount":3,"residentsCount":3,"totalArea":873.70,"consumption":{"1":1249,"2":925,"3":1449,"4":549,"6":3262,"7":7372,"8":11944,"9":6049,"10":661,"11":3349,"12":1194}},
    {"accountId":1973,"isCommercial":true,"address":"Краснодарский край, г Геленджик, с Архипо-Осиповка, ул Пограничная, д. 21 ","buildingType":"Частный","roomsCount":15,"residentsCount":3,"totalArea":689.80,"consumption":{"1":1360,"2":1870,"3":1781,"4":1806,"5":1527,"6":1193,"7":3169,"8":9361,"9":10567,"10":7128,"11":2662,"12":1184}},
    {"accountId":1976,"isCommercial":true,"address":"Краснодарский край, г Геленджик, с Архипо-Осиповка, ул Пограничная, д. 29 ","buildingType":"Частный","roomsCount":3,"residentsCount":1,"consumption":{"5":380,"6":1827,"7":447,"8":2100,"9":630,"10":870}},
    {"accountId":1979,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, село Витязево, пр-кт Южный, д. 9 , кв. 2 ","buildingType":"Многоквартирный","roomsCount":3,"residentsCount":3,"consumption":{"1":3832,"2":2106,"3":1865,"5":1981,"6":3448,"7":5182,"8":7172,"9":4494,"10":956,"12":3505}},
    {"accountId":1983,"isCommercial":true,"address":"Краснодарский край, г Геленджик, с Архипо-Осиповка, ул Ленина, д. 96 А","buildingType":"Частный","roomsCount":3,"residentsCount":3,"consumption":{"1":145,"2":213,"3":213,"4":213,"5":213,"6":0,"7":0,"8":0}},
    {"accountId":1991,"isCommercial":true,"address":"Краснодарский край, г Геленджик, с Архипо-Осиповка, ул Садовая, д. 26 ","buildingType":"Частный","roomsCount":5,"residentsCount":4,"totalArea":206.00,"consumption":{"1":1732,"2":1225,"3":1200,"4":711,"5":1594,"6":2535,"9":8072,"10":1236,"11":2119,"12":2950}},
    {"accountId":2046,"isCommercial":false,"address":"Краснодарский край, г Сочи, пгт Красная Поляна, ул Турчинского, д. 93 ","buildingType":"Частный","roomsCount":3,"residentsCount":2,"totalArea":86.70,"consumption":{"1":208,"2":2890,"3":193,"4":238,"5":188,"6":183,"7":138,"8":146,"9":125,"10":158,"11":167,"12":112}},
    {"accountId":2059,"isCommercial":false,"address":"Краснодарский край, г Сочи, с Липники, ул Липецкая, д. 8 в","buildingType":"Частный","roomsCount":2,"residentsCount":3,"totalArea":3500.00,"consumption":{"1":1036,"2":1429,"3":1483,"4":1146,"5":1131,"6":454,"8":242,"9":282,"10":327,"11":694,"12":451}},
    {"accountId":2076,"isCommercial":true,"address":"Краснодарский край, р-н Павловский, ст-ца Атаманская, ул Вокзальная, д. 74 А","buildingType":"Частный","roomsCount":3,"consumption":{"1":4399,"2":4400,"3":3995,"4":2334,"5":6174,"6":3271,"7":3272,"8":8547,"9":8548,"10":5060,"11":1359,"12":3329}},
    {"accountId":2096,"isCommercial":false,"address":"Краснодарский край, г Сочи, с Высокое, ул Афипская, д. 4 ","buildingType":"Частный","roomsCount":3,"residentsCount":3,"totalArea":179.40,"consumption":{"1":318,"2":305,"3":250,"4":850,"5":388,"6":459,"7":393,"8":434,"9":373,"10":319,"11":260,"12":175}},
    {"accountId":2185,"isCommercial":true,"address":"Краснодарский край, р-н Тимашевский, ст-ца Роговская, ул Свободная, д. 14 ","buildingType":"Частный","roomsCount":2,"residentsCount":3,"totalArea":46.90,"consumption":{"1":2704,"2":3865,"3":4934,"4":1900,"5":3328,"6":942,"7":4746,"8":5503,"9":3618,"10":2177,"11":4201,"12":2808}},
    {"accountId":2192,"isCommercial":true,"address":"Краснодарский край, р-н Темрюкский, ст-ца Тамань, ул Лебедева, д. 90 В","buildingType":"Частный","roomsCount":3,"residentsCount":1,"totalArea":1154.70,"consumption":{"1":2189,"2":2570,"3":2889,"4":1678,"5":2319,"6":4572,"7":7068,"8":5799,"9":3173,"10":1013,"11":1260,"12":2047}},
    {"accountId":2245,"isCommercial":true,"address":"Краснодарский край, р-н Анапский, село Витязево, ул Жукова, д. 58 ","buildingType":"Прочий","consumption":{"1":864,"2":781,"3":864,"4":837,"5":864,"6":4032,"7":1509,"8":1104,"9":837,"10":864,"11":837,"12":864}},
    {"accountId":2279,"isCommercial":false,"address":"Краснодарский край, г Сочи, с Альтмец, ул Кленовая, д. 19 А 1","buildingType":"Частный","roomsCount":1,"residentsCount":1,"totalArea":144.00,"consumption":{"1":345,"2":350,"3":332,"4":357,"5":333,"6":142,"7":123,"8":102,"9":136,"10":292,"11":493,"12":292}}
]""")

# Создание DataFrame
df = pd.DataFrame(data)
df['consumption'] = df['consumption'].apply(lambda x: x if isinstance(x, dict) else {})

# Функция для создания признаков
def create_features(consumption):
    months = [str(i) for i in range(1, 13)]
    values = [consumption.get(m, 0) for m in months]
    summer_months = [5, 6, 7, 8, 9]
    winter_months = [1, 2, 3, 11, 12]
    summer_values = [consumption.get(str(m), 0) for m in summer_months]
    winter_values = [consumption.get(str(m), 0) for m in winter_months]
    
    return {
        'total_consumption': sum(values),
        'avg_consumption': np.mean(values) if values else 0,
        'std_consumption': np.std(values) if len(values) > 1 else 0,
        'max_consumption': max(values) if values else 0,
        'min_consumption': min(values) if values else 0,
        'zero_months': sum(1 for v in values if v == 0),
        'summer_avg': np.mean(summer_values) if summer_values else 0,
        'winter_avg': np.mean(winter_values) if winter_values else 0,
        'max_min_diff': max(values) - min(values) if values else 0,
        'cv_consumption': np.std(values) / np.mean(values) if np.mean(values) > 0 else 0,
        'missing_months': sum(1 for m in months if m not in consumption)
    }

# Создание признаков потребления
consumption_features = pd.DataFrame(df['consumption'].apply(create_features).tolist())
df = pd.concat([df, consumption_features], axis=1)

# Обработка пропусков
for col in ['residentsCount', 'totalArea', 'roomsCount']:
    df[col] = df[col].fillna(df[col].median())

# Производные признаки
df['consumption_per_resident'] = df['total_consumption'] / df['residentsCount'].replace(0, 1)
df['consumption_per_area'] = df['total_consumption'] / df['totalArea'].replace(0, 1)
df['consumption_per_room'] = df['total_consumption'] / df['roomsCount'].replace(0, 1)
df['summer_winter_ratio'] = df['summer_avg'] / df['winter_avg'].replace(0, 1)

# Кодирование категориальных признаков
df = pd.get_dummies(df, columns=['buildingType'], prefix=['bType'], drop_first=True)

# Признаки для модели
features = [
    'total_consumption', 'std_consumption', 'max_consumption', 'zero_months',
    'summer_avg', 'winter_avg', 'max_min_diff', 'cv_consumption', 'missing_months',
    'consumption_per_resident', 'consumption_per_area', 'consumption_per_room',
    'summer_winter_ratio', 'bType_Частный', 'bType_Прочий'
]

# Целевая переменная
target = 'isCommercial'

# Проверка дисбаланса классов
print("Распределение классов до балансировки:")
print(df[target].value_counts())

# Борьба с дисбалансом классов
X = df[features].fillna(0)
y = df[target]

try:
    # Попытка использовать SMOTE с уменьшенным количеством соседей
    smote = SMOTE(
        random_state=42,
        k_neighbors=2,
        sampling_strategy='auto'
    )
    X_res, y_res = smote.fit_resample(X, y)
    print("\nИспользован SMOTE для балансировки")
except ValueError as e:
    print(f"\nОшибка SMOTE: {e}")
    print("Используем RandomOverSampler вместо SMOTE")
    ros = RandomOverSampler(random_state=42)
    X_res, y_res = ros.fit_resample(X, y)

print("\nРаспределение классов после балансировки:")
print(pd.Series(y_res).value_counts())

# Разделение данных
X_train, X_test, y_train, y_test = train_test_split(
    X_res, y_res, test_size=0.2, random_state=42
)

# Нормализация данных
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Обучение модели
model = RandomForestClassifier(
    n_estimators=200,
    max_depth=8,
    class_weight='balanced',
    random_state=42
)
model.fit(X_train_scaled, y_train)

# Прогнозирование и оценка
y_pred = model.predict(X_test_scaled)
print("\nОтчёт о классификации:")
print(classification_report(y_test, y_pred))

print("\nМатрица ошибок:")
print(confusion_matrix(y_test, y_pred))

# Важность признаков
feature_importance = pd.DataFrame({
    'Признак': features,
    'Важность': model.feature_importances_
}).sort_values('Важность', ascending=False)

print("\nТоп-5 важных признаков:")
print(feature_importance.head(5))

# Визуализация важности признаков
fig = px.bar(feature_importance.sort_values('Важность'), 
             x='Важность', y='Признак', 
             title='Важность признаков для определения коммерции')
fig.show()

# Прогнозирование для всех данных
df['predicted_commercial'] = model.predict(scaler.transform(X.fillna(0)))

# Визуализация результатов
fig = px.scatter(
    df,
    x='total_consumption',
    y='consumption_per_area',
    color='predicted_commercial',
    color_discrete_map={True: '#FFD700', False: '#0066CC'},
    size='totalArea',
    hover_data=['accountId', 'address', 'roomsCount'],
    title='Прогноз коммерческих аккаунтов',
    labels={
        'total_consumption': 'Общее потребление (кВт·ч)',
        'consumption_per_area': 'Потребление на м²',
        'predicted_commercial': 'Прогноз коммерции'
    }
)
fig.update_layout(legend_title_text='Коммерческий аккаунт')
fig.show()

# Сохранение результатов
results = df[['accountId', 'address', 'predicted_commercial']]
results.to_csv('commercial_predictions.csv', index=False)
files.download('commercial_predictions.csv')

print("\nГотово! Результаты сохранены в файле commercial_predictions.csv")
